File: packages/typescript/src/interpreters/socket.ts
/**
 * Socket inbound: a held-open connection that feeds RawInbound to a sink.
 *
 * `WebSocket` is a global in Bun and Node 22+. The event-based API is wrapped
 * into awaitable `receive()` so the session reads as an ordinary loop, and so
 * tests can inject a fake without a network.
 *
 * Frames that arrive while nothing is awaiting are queued, never dropped.
 */

import * as Effect from "effect/Effect"
import { socketDriver, type SocketDriver } from "../adapters/socket.ts"
import type { Transport } from "./transport.ts"

export { socketDriver }

export type GatewaySocket = {
  /** Resolves with the next frame, or rejects when the socket closes. */
  readonly receive: () => Promise<string>
  readonly send: (data: string) => Promise<void>
  readonly close: () => void
}

export type SocketOpener = (url: string) => Promise<GatewaySocket>

export type SocketSink = (
  body: unknown,
  headers?: { readonly [key: string]: string },
) => Effect.Effect<ReadonlyArray<unknown>>

/** Fatal: retrying cannot help (bad token). Stop the loop. */
export class SocketFatal extends Error {
  readonly _tag = "SocketFatal"
}

/** Opens a real WebSocket and adapts it to GatewaySocket. */
export const openWebSocket: SocketOpener = (url) =>
  new Promise<GatewaySocket>((resolveOpen, rejectOpen) => {
    const socket = new WebSocket(url)
    const queued: string[] = []
    const waiting: Array<{
      resolve: (value: string) => void
      reject: (error: Error) => void
    }> = []
    let closed: Error | undefined

    const deliver = (frame: string): void => {
      const next = waiting.shift()
      if (next === undefined) {
        queued.push(frame)
        return
      }
      next.resolve(frame)
    }

    const fail = (error: Error): void => {
      closed = error
      while (waiting.length > 0) {
        waiting.shift()?.reject(error)
      }
    }

    socket.addEventListener("message", (event: MessageEvent) => {
      deliver(typeof event.data === "string" ? event.data : String(event.data))
    })
    socket.addEventListener("close", () => fail(new Error("socket closed")))
    socket.addEventListener("error", () => fail(new Error("socket error")))
    socket.addEventListener("open", () =>
      resolveOpen({
        receive: () =>
          new Promise<string>((resolve, reject) => {
            const buffered = queued.shift()
            if (buffered !== undefined) {
              resolve(buffered)
              return
            }
            if (closed !== undefined) {
              reject(closed)
              return
            }
            waiting.push({ resolve, reject })
          }),
        send: async (data: string) => {
          socket.send(data)
        },
        close: () => socket.close(),
      }),
    )
    socket.addEventListener("error", () => rejectOpen(new Error("socket failed to open")))
  })

/** Sleep, used for reconnect backoff. */
export const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms))

const asRecord = (value: unknown): { readonly [key: string]: unknown } =>
  typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as { readonly [key: string]: unknown })
    : {}

export class SocketSession {
  #eventsSeen = 0
  #stop = false

  constructor(
    readonly driver: SocketDriver,
    readonly sink: SocketSink,
    readonly options: {
      readonly transport: Transport
      readonly open?: SocketOpener
      readonly log?: (message: string) => void
    },
  ) {}

  stop(): void {
    this.#stop = true
  }

  /** Tests restart a session after maxEvents; not part of the public SDK. */
  restart(): void {
    this.#stop = false
  }

  run(
    options: { readonly maxEvents?: number } = {},
  ): Effect.Effect<ReadonlyArray<unknown>> {
    return Effect.promise(async () => {
      this.#stop = false
      const collected: unknown[] = []
      let backoff = 1000
      while (!this.#stop) {
        try {
          await this.#runOnce(options.maxEvents, collected)
          backoff = 1000
        } catch (error) {
          if (error instanceof SocketFatal) {
            this.options.log?.(`fatal socket error, not retrying: ${String(error)}`)
            return collected
          }
          if (this.#stop) {
            break
          }
          this.options.log?.(
            `socket dropped (${String(error)}); reconnecting in ${backoff / 1000}s`,
          )
          await delay(backoff)
          backoff = Math.min(backoff * 2, 60_000)
        }
      }
      return collected
    })
  }

  async #runOnce(maxEvents: number | undefined, collected: unknown[]): Promise<void> {
    const dispatched = await Effect.runPromise(
      this.options.transport.dispatch(this.driver.openPlan()),
    )
    const opened = this.driver.urlOf(dispatched)
    if (opened.fatal !== undefined && opened.fatal.length > 0) {
      throw new SocketFatal(opened.fatal)
    }
    const url = opened.url ?? ""
    if (url === "") {
      throw new Error("no socket url")
    }
    const open = this.options.open ?? openWebSocket
    const socket = await open(url)
    let cancelHeartbeat: (() => void) | undefined
    try {
      while (!this.#stop) {
        let raw: string
        try {
          raw = await socket.receive()
        } catch {
          throw new Error("socket closed")
        }
        let frame: { readonly [key: string]: unknown }
        try {
          frame = asRecord(JSON.parse(raw) as unknown)
        } catch {
          continue
        }
        const decision = this.driver.onFrame(frame)
        for (const payload of decision.send ?? []) {
          await socket.send(payload)
        }
        if (decision.heartbeatInterval !== undefined) {
          cancelHeartbeat?.()
          cancelHeartbeat = this.#startHeartbeat(socket, decision.heartbeatInterval)
        }
        if (decision.fatal !== undefined && decision.fatal.length > 0) {
          throw new SocketFatal(decision.fatal)
        }
        if (decision.sink !== undefined) {
          this.#eventsSeen += 1
          collected.push(...(await Effect.runPromise(this.sink(decision.sink, {}))))
          if (maxEvents !== undefined && this.#eventsSeen >= maxEvents) {
            this.#stop = true
            return
          }
        }
        if (decision.reconnect === true) {
          return
        }
      }
    } finally {
      cancelHeartbeat?.()
      socket.close()
    }
  }

  #startHeartbeat(socket: GatewaySocket, intervalSec: number): () => void {
    let live = true
    const beat = async (): Promise<void> => {
      await delay(intervalSec * 1000 * Math.random())
      while (live && !this.#stop) {
        try {
          const payload = this.driver.heartbeatPayload()
          if (payload !== undefined) {
            await socket.send(payload)
          }
        } catch {
          return
        }
        await delay(intervalSec * 1000)
      }
    }
    void beat()
    return () => {
      live = false
    }
  }
}

File: packages/python/src/caspian/interpreters/socket.py
"""Socket session — the one inbound loop for held-open connections.

Webhook, poll, and socket all end at ProcessInterpreter.handle_webhook. This
module is how bytes arrive over a socket: connect, recv, backoff. The driver
(an adapter) unwraps frames. asyncio/websockets live here, never in core.
`websockets` is optional: caspian[discord] or caspian[slack-socket].
"""

from __future__ import annotations

import asyncio
import json
import logging
import random
from collections.abc import Callable
from typing import Any

from caspian.adapters.socket import SocketDriver
from caspian.core.ports import RawInbound, Result, TransportPort

Sink = Callable[[RawInbound], list[Result]]

log = logging.getLogger("caspian.socket")


class SocketFatal(Exception):
    """Driver says retrying cannot help (bad token). Stop the loop."""


class SocketSession:
    """Hold one socket open and feed inbound to a sink."""

    def __init__(
        self,
        driver: SocketDriver,
        sink: Sink,
        *,
        transport: TransportPort,
        connect: Any = None,
    ) -> None:
        self._driver = driver
        self._sink = sink
        self._transport = transport
        self._connect = connect
        self._results: list[Result] = []
        self._events_seen = 0
        self._max_events: int | None = None
        self._stop = False

    def stop(self) -> None:
        self._stop = True

    async def run(self, *, max_events: int | None = None) -> list[Result]:
        """Hold the connection open, reconnecting forever. Returns sink results."""
        self._max_events = max_events
        self._stop = False
        backoff = 1.0
        while not self._stop:
            try:
                await self._run_once()
                backoff = 1.0
            except SocketFatal as exc:
                log.error("fatal socket error, not retrying: %s", exc)
                return self._results
            except Exception as exc:  # noqa: BLE001 - any drop is a reconnect
                if self._stop:
                    break
                log.warning("socket dropped (%s); reconnecting in %.0fs", exc, backoff)
                await asyncio.sleep(backoff)
                backoff = min(backoff * 2, 60)
        return self._results

    async def _run_once(self) -> None:
        planned = self._driver.open_plan()
        if not planned.is_ok:
            reason = planned.error.reason if planned.error is not None else "open failed"
            raise SocketFatal(reason)
        dispatched = self._transport.dispatch(planned.value)
        if not dispatched.is_ok:
            err = dispatched.error
            reason = err.reason if err is not None else "open dispatch failed"
            raise ConnectionError(reason)
        opened = self._driver.url_of(dispatched.value)
        if opened.fatal:
            raise SocketFatal(opened.fatal)
        if not opened.url:
            raise ConnectionError("no socket url")
        connect = self._connect
        if connect is None:
            import websockets  # type: ignore[import-not-found]

            connect = websockets.connect
        heartbeat: asyncio.Task[None] | None = None
        async with connect(opened.url, **self._driver.connect_kwargs()) as ws:
            try:
                while not self._stop:
                    try:
                        raw = await ws.recv()
                    except (ValueError, TypeError):
                        continue
                    try:
                        frame = json.loads(raw)
                    except (json.JSONDecodeError, TypeError, ValueError):
                        continue
                    if not isinstance(frame, dict):
                        continue
                    decision = self._driver.on_frame(frame)
                    for payload in decision.send:
                        await ws.send(payload)
                    if decision.heartbeat_interval is not None:
                        if heartbeat is not None:
                            heartbeat.cancel()
                        heartbeat = asyncio.create_task(
                            self._heartbeat(ws, decision.heartbeat_interval)
                        )
                    if decision.fatal:
                        raise SocketFatal(decision.fatal)
                    if decision.sink is not None:
                        self._events_seen += 1
                        self._results.extend(
                            await asyncio.to_thread(self._sink, decision.sink)
                        )
                        if (
                            self._max_events is not None
                            and self._events_seen >= self._max_events
                        ):
                            self._stop = True
                            return
                    if decision.reconnect:
                        return
            finally:
                if heartbeat is not None:
                    heartbeat.cancel()

    async def _heartbeat(self, ws: Any, interval: float) -> None:
        await asyncio.sleep(interval * random.random())
        while True:
            payload = self._driver.heartbeat_payload()
            if payload:
                await ws.send(payload)
            await asyncio.sleep(interval)


__all__ = ["Sink", "SocketFatal", "SocketSession"]
